src/store/mod.rs

```diff
@@ -0,0 +1,147 @@
+use crate::node_pages_map::{Page, PAGE_SIZE};
+use io_uring::{cqueue, opcode, squeue, types, IoUring};
+use slab::Slab;
+use std::{
+    fs::{File, OpenOptions},
+    io,
+    ops::Range,
+    os::{fd::AsRawFd, unix::fs::OpenOptionsExt},
+    path::PathBuf,
+    sync::Arc,
+};
+
+#[cfg(test)]
+mod tests;
+
+const RING_CAPACITY: u32 = 128;
+
+/// The Store is an on disk array of [`crate::node_pages_map::Page`]
+pub struct Store {
+    store_file: File,
+}
+
+/// Options to open a Store
+pub struct StoreOptions {
+    /// File path of the storage file.
+    file_path: PathBuf,
+}
+
+impl Default for StoreOptions {
+    fn default() -> Self {
+        Self {
+            file_path: PathBuf::from("node_pages_store"),
+        }
+    }
+}
+
+impl Store {
+    /// Create a new Store given the StoreOptions
+    pub fn new(options: StoreOptions) -> Self {
+        let store_file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .custom_flags(libc::O_DIRECT)
+            .open(options.file_path)
+            .expect("could not open file");
+
+        Store { store_file }
+    }
+}
+
+/// An I/O handle for a store. This owns an io_uring ring which is used to submit reads and writes.
+pub struct StoreIo {
+    store: Arc<Store>,
+    ring: IoUring,
+    pending: Slab<Pending>,
+}
+
+impl StoreIo {
+    pub fn new(store: Arc<Store>) -> Self {
+        StoreIo {
+            ring: IoUring::<squeue::Entry, cqueue::Entry>::builder()
+                .setup_sqpoll(100)
+                // .setup_iopoll
+                .build(RING_CAPACITY)
+                .expect("Error building io_uring"),
+            store,
+            pending: Slab::with_capacity(RING_CAPACITY as _),
+        }
+    }
+
+    pub fn is_submit_full(&mut self) -> bool {
+        self.ring.submission().is_full()
+    }
+
+    // read a page. panics if the submission queue is full.
+    pub fn read(&mut self, page_id: u64, mut buf: Box<Page>) -> io::Result<()> {
+        let entry = opcode::Read::new(
+            types::Fd(self.store.store_file.as_raw_fd()),
+            &mut buf[0] as *mut u8,
+            PAGE_SIZE as u32,
+        )
+        .offset(page_id * (PAGE_SIZE as u64 + 1))
+        .build()
+        .user_data(self.pending.vacant_key() as u64);
+        unsafe { self.ring.submission().push(&entry).unwrap() }
+        self.pending.insert(Pending {
+            kind: Kind::Read,
+            page_id,
+            buf,
+        });
+
+        self.ring.submit()?;
+        Ok(())
+    }
+
+    // write a page. panics if the submission queue is full.
+    pub fn write<'a>(&mut self, page_id: u64, buf: Box<Page>) -> io::Result<()> {
+        let entry = opcode::Write::new(
+            types::Fd(self.store.store_file.as_raw_fd()),
+            &buf[0] as *const u8,
+            PAGE_SIZE as u32,
+        )
+        .offset(page_id * (PAGE_SIZE as u64 + 1))
+        .build()
+        .user_data(self.pending.vacant_key() as u64);
```

Comment:
nit: maybe you'd better insert pending first and get the id. That's the idiomatic way of using slab

Reply:
It works better not to in this case, as we would have to either keep the pointer or reach inside the slab to borrow it if we insert it first

Reply:
eh? not sure if I follow. I had in mind something like this:

1. save the buf ptr.
2. insert `Pending` into slab. slab returns a 'static id (u64 or usize, IDR)
3. we create a new Write SQE passing the *const u8 buf pointer and the ID returned by the slab.